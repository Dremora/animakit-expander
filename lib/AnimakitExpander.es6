import React                        from 'react';
import { findDOMNode }              from 'react-dom';
import { isEqual, genUniqueString } from 'animakit-core';

export default class AnimakitExpander extends React.Component {
  static propTypes = {
    children:   React.PropTypes.any,
    expanded:   React.PropTypes.bool,
    horizontal: React.PropTypes.bool,
    align:      React.PropTypes.string,
    duration:   React.PropTypes.number,
    easing:     React.PropTypes.string
  };

  static defaultProps = {
    expanded:   true,
    horizontal: false,
    align:      'left',
    duration:   500,
    easing:     'ease-out'
  };

  state = {
    animation: false,
    expanded:  false
  };

  styleSheetNode   = null;
  animationNames   = null;
  contentNode      = null;
  contentSize      = 0;
  animationResetTO = null;

  componentDidMount() {
    this.contentNode = findDOMNode(this.refs.content);
    this.repaint(this.props, true);
  }

  componentWillReceiveProps(nextProps) {
    this.repaint(nextProps);
  }

  shouldComponentUpdate(nextProps, nextState) {
    const stateChanged = !isEqual(nextState, this.state);

    const propsChanged = !isEqual(nextProps.children, this.props.children);

    return stateChanged || propsChanged;
  }

  componentWillUnmount() {
    this.cancelAnimationReset();
  }

  startAnimationReset() {
    this.animationResetTO = setTimeout(() => {
      this.setState({
        animation: false
      });
    }, this.props.duration);
  }

  cancelAnimationReset() {
    if (this.animationResetTO) clearTimeout(this.animationResetTO);
  }

  checkResize() {
    const size = this.calcSize();
    if (size !== this.contentSize) {
      this.contentSize = size;
      this.resetAnimationNames();
      this.resetAnimationStyles();
    }
  }

  calcSize() {
    const node = this.contentNode;
    return this.props.horizontal ? node.offsetWidth : node.offsetHeight;
  }

  resetDimensionsState(stateChunk) {
    const { width, height } = stateChunk;

    if (
      width === this.state.width &&
      height === this.state.height
    ) return {};

    return stateChunk;
  }

  repaint(nextProps, first = false) {
    const expanded = nextProps.expanded;

    if (this.state.expanded === expanded) return;

    const animation = !first;
    const state = { expanded, animation };

    if (animation) {
      this.checkResize();
      this.cancelAnimationReset();
    }

    this.setState(state);

    if (animation) {
      this.startAnimationReset();
    }
  }

  resetAnimationNames() {
    const uniqueString = genUniqueString();
    const expand = `expand-${ uniqueString }`;
    const collapse = `collapse-${ uniqueString }`;

    this.animationNames = { expand, collapse };
  }

  getAnimationStyle(type) {
    const name = this.animationNames[type];
    const dimension = this.props.horizontal ? 'width' : 'height';
    const size = this.contentSize;
    const from = type === 'expand' ? 0 : size;
    const to = type === 'expand' ? size : 0;

    return `@keyframes ${ name } { from { ${ dimension }: ${ from }px; } to { ${ dimension }: ${ to }px; } }`;
  }

  resetAnimationStyles() {
    if (!this.styleSheetNode) {
      const sheet = document.createElement('style');
      sheet.setAttribute('type', 'text/css');
      document.head.appendChild(sheet);

      this.styleSheetNode = sheet;
    }

    this.styleSheetNode.innerHTML = `${ this.getAnimationStyle('expand') } ${ this.getAnimationStyle('collapse') }`;
  }

  getWrapperStyles() {
    const expanded = this.state.expanded;
    const position = 'relative';
    const overflow = 'hidden';

    if (!this.state.animation) {
      const horizontal = this.props.horizontal;

      if (!expanded) {
        const width = 0;
        const height = 0;

        return horizontal ? { overflow, width } : { overflow, height };
      }

      const width = 'auto';
      const height = 'auto';

      return horizontal ? { width } : { height };
    }

    const { duration, easing } = this.props;
    const name = this.animationNames[expanded ? 'expand' : 'collapse'];

    const animation = `${ name } ${ duration }ms ${ easing } forwards`;

    return { position, overflow, animation };
  }

  getContentStyles() {
    if (this.props.horizontal) {
      const float = this.props.align === 'right' ? 'right' : 'left';

      return { float };
    }

    if (this.props.align === 'bottom' && this.state.animation) {
      return {
        position: 'absolute',
        bottom:   0,
        left:     0,
        width:    '100%'
      };
    }

    return {};
  }

  render() {
    return (
      <div>
        <div style = { this.getWrapperStyles() }>
          <div
            ref = "content"
            style = { this.getContentStyles() }
          >
            <span style = {{ display: 'table', height: 0 }}></span>
            { this.props.children }
          </div>
        </div>
      </div>
    );
  }
}
