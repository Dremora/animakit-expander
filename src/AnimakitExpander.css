.root {
  all: initial;
}
.wrapper {

}
.content {

}
.content::before {
  content: '';
  display: table;
  height: 0;
}
