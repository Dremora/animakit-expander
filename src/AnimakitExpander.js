import React                        from 'react';
import { findDOMNode }              from 'react-dom';

import styles                       from './styles.js';
import useSheet                     from 'react-jss';
import jss                          from 'jss';
import nested                       from 'jss-nested';

import { isEqual, genUniqueString } from 'animakit-core';

jss.use(nested());

class AnimakitExpander extends React.Component {
  static propTypes = {
    children:   React.PropTypes.any,
    expanded:   React.PropTypes.bool,
    horizontal: React.PropTypes.bool,
    align:      React.PropTypes.string,
    duration:   React.PropTypes.number,
    easing:     React.PropTypes.string
  };

  static defaultProps = {
    expanded:   true,
    horizontal: false,
    align:      'left',
    duration:   500,
    easing:     'ease-out'
  };

  state = {
    animation: false,
    expanded:  false
  };

  styleSheetNode   = null;
  animationNames   = null;
  contentNode      = null;
  contentSize      = 0;
  animationResetTO = null;

  componentWillMount() {
    this.styleSheetNode = document.createElement('style');
    document.head.appendChild(this.styleSheetNode);
  }

  componentDidMount() {
    this.contentNode = findDOMNode(this.refs.content);
    this.repaint(this.props, true);
  }

  componentWillReceiveProps(nextProps) {
    this.repaint(nextProps);
  }

  shouldComponentUpdate(nextProps, nextState) {
    const stateChanged = !isEqual(this.state, nextState);

    return stateChanged;
  }

  componentWillUnmount() {
    this.cancelAnimationReset();
  }

  startAnimationReset() {
    this.animationResetTO = setTimeout(() => {
      this.setState({
        animation: false
      });
    }, this.props.duration);
  }

  cancelAnimationReset() {
    if (this.animationResetTO) clearTimeout(this.animationResetTO);
  }

  checkResize() {
    const size = this.calcSize();
    if (size !== this.contentSize) {
      this.contentSize = size;
      this.resetAnimationNames();
      this.resetAnimationStyles();
    }
  }

  calcSize() {
    const node = this.contentNode;
    return this.props.horizontal ? node.offsetWidth : node.offsetHeight;
  }

  resetDimensionsState(stateChunk) {
    const { width, height } = stateChunk;

    if (
      width === this.state.width &&
      height === this.state.height
    ) return {};

    return stateChunk;
  }

  repaint(nextProps, first = false) {
    const expanded = nextProps.expanded;

    if (this.state.expanded === expanded) return;

    const animation = !first;
    const state = { expanded, animation };

    if (animation) {
      this.checkResize();
      this.cancelAnimationReset();
    }

    this.setState(state);

    if (animation) {
      this.startAnimationReset();
    }
  }

  resetAnimationNames() {
    const uniqueString = genUniqueString();
    const expand = `AnimakitExpanderExpand-${ uniqueString }`;
    const collapse = `AnimakitExpanderCollapse-${ uniqueString }`;

    this.animationNames = { expand, collapse };
  }

  getAnimationStyle(expand) {
    const name = this.animationNames[expand ? 'expand' : 'collapse'];
    const dimension = this.props.horizontal ? 'width' : 'height';
    const size = this.contentSize;
    const from = expand ? 0 : size;
    const to = expand ? size : 0;

    return `@keyframes ${ name } { from { ${ dimension }: ${ from }px; } to { ${ dimension }: ${ to }px; } }`;
  }

  resetAnimationStyles() {
    this.styleSheetNode.innerHTML = `${ this.getAnimationStyle(true) } ${ this.getAnimationStyle(false) }`;
  }

  getWrapperStyles() {
    const expanded = this.state.expanded;
    const position = 'relative';
    const overflow = 'hidden';

    if (!this.state.animation) {
      const horizontal = this.props.horizontal;

      if (!expanded) {
        const width = 0;
        const height = 0;

        return horizontal ? { overflow, width } : { overflow, height };
      }

      const width = 'auto';
      const height = 'auto';

      return horizontal ? { width } : { height };
    }

    const { duration, easing } = this.props;
    const name = this.animationNames[expanded ? 'expand' : 'collapse'];

    const animation = `${ name } ${ duration }ms ${ easing } forwards`;

    return { position, overflow, animation };
  }

  getContentStyles() {
    if (this.props.horizontal) {
      const float = this.props.align === 'right' ? 'right' : 'left';

      return { float };
    }

    if (this.props.align === 'bottom' && this.state.animation) {
      return {
        position: 'absolute',
        bottom:   0,
        left:     0,
        width:    '100%'
      };
    }

    return {};
  }

  render() {
    const { classes } = this.props.sheet;

    return (
      <div className = { classes.root }>
        <div
          className = { classes.wrapper }
          style = { this.getWrapperStyles() }
        >
          <div
            ref = "content"
            className = { classes.content }
            style = { this.getContentStyles() }
          >
            { this.props.children }
          </div>
        </div>
      </div>
    );
  }
}

export default useSheet(AnimakitExpander, styles);
export { AnimakitExpander as PureAnimakitExpander };
