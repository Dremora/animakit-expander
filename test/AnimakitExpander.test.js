/* eslint-env jest */

import { shallow }      from 'enzyme';

import React            from 'react';
import AnimakitExpander from '../lib/AnimakitExpander.js';

describe('<AnimakitExpander />', () => {
  it('should render', () => {
    const component = shallow(<AnimakitExpander />);

    expect(component).toMatchSnapshot();
  });
});
