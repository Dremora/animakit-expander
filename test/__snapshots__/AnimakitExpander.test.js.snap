// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimakitExpander /> should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    style={Object {}}
>
    <div
        style={Object {}}
    >
        
    </div>
</div>,
  "nodes": Array [
    <div
      style={Object {}}
>
      <div
            style={Object {}}
      >
            
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <AnimakitBase
        align="left"
        duration={500}
        durationPerPx={0}
        easing="ease-out"
        expanded={true}
        horizontal={false}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": AnimakitExpander {
        "_reactInternalInstance": [Circular],
        "animationResetTO": null,
        "canAnimate": false,
        "changingProps": Array [],
        "contentNode": null,
        "context": Object {},
        "listeners": Object {
          "checkResize": [Function],
        },
        "props": Object {
          "align": "left",
          "duration": 500,
          "durationPerPx": 0,
          "easing": "ease-out",
          "expanded": true,
          "horizontal": false,
        },
        "refs": Object {},
        "resizeCheckerRAF": null,
        "state": Object {
          "animation": false,
          "duration": 0,
          "expanded": false,
          "lastSize": 0,
          "prepare": false,
          "size": 0,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "useWinResize": false,
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          style={Object {}}
>
          <div
                    style={Object {}}
          >
                    
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          style={Object {}}
>
          <div
                    style={Object {}}
          >
                    
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <AnimakitBase
    align="left"
    duration={500}
    durationPerPx={0}
    easing="ease-out"
    expanded={true}
    horizontal={false}
/>,
}
`;
